drivers/dac/dac_dac161s997.c

```diff
@@ -0,0 +1,311 @@
+	k_sem_give(&data->lock);
+
+	return ret;
+}
+
+static int dac161s997_read_status(const struct device *dev, union dac161s997_status *status)
+{
+	int ret;
+	uint16_t tmp;
+
+	ret = dac161s997_read_reg(dev, DAC161S997_REG_STATUS, &tmp);
+	if (ret == 0) {
+		status->raw = tmp;
+	}
+
+	return ret;
+}
+
+static void dac161s997_gpio_errb_work_handler(struct k_work *work)
+{
+	struct dac161s997_data *data = CONTAINER_OF(work, struct dac161s997_data, gpio_errb_work);
+	const struct device *dev = data->dev;
+	union dac161s997_status status;
+	int ret;
+
+	ret = k_sem_take(&data->lock, K_FOREVER);
+	if (ret != 0) {
+		LOG_WRN("ERRB handler take lock failed: %d", ret);
+		return;
+	}
+
+	ret = dac161s997_read_status(dev, &status);
+
+	if (data->error_cb != NULL) {
+		data->error_cb(dev, ret == 0 ? &status : NULL);
+	}
+
+	k_sem_give(&data->lock);
+}
+
+static void dac161s997_gpio_errb_cb(const struct device *dev, struct gpio_callback *cb,
+				    gpio_port_pins_t pins)
+{
+	struct dac161s997_data *data = CONTAINER_OF(cb, struct dac161s997_data, gpio_errb_cb);
+	int ret;
+
+	ret = k_work_submit(&data->gpio_errb_work);
+	if (ret != 1) {
+		LOG_WRN("ERRB work not queued: %d", ret);
+	}
+}
+
+static int dac161s997_init(const struct device *dev)
+{
+	const struct dac161s997_config *config = dev->config;
+	struct dac161s997_data *data = dev->data;
+	union dac161s997_status status;
+	int ret;
+
+	data->dev = dev;
+
+	if (!spi_is_ready_dt(&config->bus)) {
+		LOG_ERR("SPI bus %s not ready", config->bus.bus->name);
+		return -ENODEV;
+	}
+
+	k_sem_init(&data->lock, 0, 1);
+
+	ret = dac161s997_write_reg(dev, DAC161S997_REG_RESET, 0xc33c);
+	if (ret != 0) {
+		return ret;
+	}
+	ret = dac161s997_write_reg(dev, DAC161S997_REG_NOP, 0);
+	if (ret != 0) {
+		return ret;
+	}
+
+	/* Read status to clear any sticky error caused during boot or reboot */
+	ret = dac161s997_read_status(dev, &status);
+	if (ret != 0) {
+		return ret;
+	}
+
+	/* Check that DAC_RES bits are all set */
+	if (status.dac_resolution != 0x7) {
+		LOG_ERR("Unexpected DAC resolution value: 0x%02x", status.dac_resolution);
+		return ret;
+	}
+
+	if (config->gpio_errb.port != NULL) {
+		if (!gpio_is_ready_dt(&config->gpio_errb)) {
+			LOG_ERR("ERRB GPIO is not ready");
+			return -ENODEV;
+		}
+
+		k_work_init(&data->gpio_errb_work, dac161s997_gpio_errb_work_handler);
+
+		gpio_init_callback(&data->gpio_errb_cb, dac161s997_gpio_errb_cb,
+				   BIT(config->gpio_errb.pin));
+
+		ret = gpio_pin_configure_dt(&config->gpio_errb, GPIO_INPUT);
+		if (ret != 0) {
+			LOG_ERR("Configure ERRB GPIO failed: %d", ret);
+			return ret;
+		}
+
+		ret = gpio_pin_interrupt_configure_dt(&config->gpio_errb, GPIO_INT_EDGE_TO_ACTIVE);
+		if (ret) {
+			LOG_ERR("Configure ERRB interrupt failed: %d", ret);
+			return ret;
+		}
+
+		ret = gpio_add_callback_dt(&config->gpio_errb, &data->gpio_errb_cb);
+		if (ret != 0) {
+			LOG_ERR("Configure ERRB callback failed: %d", ret);
+			return ret;
+		}
+	}
+
+	k_sem_give(&data->lock);
```

Comment:
You could just initialise the lock accordingly further up this function?

Reply:
It should be possible since `gpio_add_callback_dt` is called lastly.